### PR TITLE
[DSIP-38] Add plugin directory in binary package

### LINKED ISSUE #16096: [Improvement] Shade plugin jar
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### LINKED ISSUE #16097: [Improvement] Move plugin to plugin directory
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by SbloodyS**
## Purpose of the pull request

close #16096
close #16097

## Brief change log

- Shade the plugins and put them into target plugins.
- Add plugins directory.
- Move each server's libs to shared libs directory and create a symbolic link from shared libs to eash server's libs.
- Add dependency `dolphinscheduler-common` provided from `dolphinscheduler-spi` and `dolphinscheduler-plugin` module.
- Refactor dist package and docker build/push logic base on the new binary package.
- Add `API-Test-Result` to mergeable check.

Then the new binary package will be like
```
- dolphinscheduler-bin
  - api-server
  - master-server
  - alert-server
  - worker-server
  - standalone-server
  - tools
  - libs 370M
  - plugins
    - task-plugins 494M
    - alert-plugins 32M
    - storage-plugins 72M
    - datasource-plugins 284M
```

### binary package size
before: 1.2G
after(without plugins directory): 450M

### Todo
- Remove duplicated jars in shade from libs directory.
- Create shade plugins to push apache nexus's repository github action and remove plugins during release package and add it to the upgrade docs.

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java [hunk 1]
```diff
@@ -68,7 +67,7 @@ public AlertResult sendMessage(AlertData alertData) {
             log.error("Send prometheus alert manager alert error", e);
             alertResult = new AlertResult();
             alertResult.setSuccess(false);
-            alertResult.setMessage(ExceptionUtils.getMessage(e));
```
Comment (ruanwenjun) [inline dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java]: Please revert this change, we need to use `ExceptionUtils` to get the exception stack.

File: pom.xml [hunk 2]
```diff
@@ -477,64 +507,6 @@
                     <groupId>org.codehaus.mojo</groupId>
                     <artifactId>exec-maven-plugin</artifactId>
                     <version>${exec-maven-plugin.version}</version>
-                    <executions>
```
Comment (ruanwenjun) [inline pom.xml:468]: Why remove this?

File: dolphinscheduler-dist/src/main/assembly/assembly-plugins.sh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+storage-plugins
+task-plugins
+)
+
+for plugin_path in ${PLUGINS_PATH[@]}
+do
+  cd $BIN_DIR/plugins/$plugin_path
+  find ./* -name "*.jar" | xargs -I {} mv {} ./
+  ls -d */ | xargs -I {} rm -rf {}
+done
```
Comment (ruanwenjun) [inline dolphinscheduler-dist/src/main/assembly/assembly-plugins.sh:44]: We should use maven assembly plugin to manage the plugin directory

File: dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/bin/start.sh [hunk 4]
```diff
@@ -18,11 +18,12 @@
 set -eo pipefail
 
 BIN_DIR=$(dirname $(readlink -f "$0"))
-DOLPHINSCHEDULER_HOME=$(cd ${BIN_DIR}/..;pwd)
+DOLPHINSCHEDULER_HOME=$(cd ${BIN_DIR}/../..;pwd)
```
Comment (ruanwenjun) [inline dolphinscheduler-alert/dolphinscheduler-alert-server/src/main/bin/start.sh:21]: Why change this value?

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java (hunk 1 above)
Comment (SbloodyS) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java]: We've already print the stack trace in L68. I think we don't need to print stack trace here. WDYT?

File: pom.xml (hunk 2 above)
Comment (SbloodyS) [reply pom.xml:468]: Move docker build logic to the `dolphinscheduler-dist` module since docker build logic refactored.

File: dolphinscheduler-dist/src/main/assembly/assembly-plugins.sh (hunk 3 above)
Comment (SbloodyS) [reply dolphinscheduler-dist/src/main/assembly/assembly-plugins.sh:44]: Assemby packaged directory:
- plugins
  - alert-plugins
    - dolphinscheduler-alert-dingtalk/taget/dolphinscheduler-alert-dingtalk-shade.jar

Directory processed with shell:
- plugins
  - alert-plugins
    - dolphinscheduler-alert-dingtalk-shade.jar

Assemby fuzzy matching directory can't remove the file path, so it needs secondary processing with shell.

File: dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java (hunk 1 above)
Comment (SbloodyS) [reply dolphinscheduler-alert/dolphinscheduler-alert-plugins/dolphinscheduler-alert-prometheus/src/main/java/org/apache/dolphinscheduler/plugin/alert/prometheus/PrometheusAlertSender.java]: Fixed.

**Review by ruanwenjun — APPROVED** [review]
LGTM

**Review by EricGao888 — APPROVED** [review]

**Review by EricGao888 — APPROVED** [review]

**Review by EricGao888 — APPROVED** [review]

**Comment by EricGao888** [comment]
Seems there were some network issues when I clicked `Approve`
